src/stdlib_math.fypp

```diff
@@ -359,6 +359,28 @@ module stdlib_math
         #:endfor
         #:endfor
     end interface all_close
+    
+    !> Version: experimental
+    !>
+    !> Computes differences between adjacent elements of the input array.
+    !> ([Specification](../page/specs/stdlib_math.html#diff))
+    interface diff
+        #:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+        #:for k1, t1 in RI_KINDS_TYPES
+        pure module function diff_1_${k1}$(x, n, prepend, append) result(Y)
+            ${t1}$, intent(in) :: x(:)
+            integer, intent(in), optional :: n
+            ${t1}$, intent(in), optional :: prepend(:), append(:)
+            ${t1}$, allocatable :: Y(:)
+        end function diff_1_${k1}$
+        pure module function diff_2_${k1}$(X, n, dim, prepend, append) result(Y)
+            ${t1}$, intent(in) :: X(:, :)
+            integer, intent(in), optional :: n, dim
+            ${t1}$, intent(in), optional :: prepend(:, :), append(:, :)
+            ${t1}$, allocatable :: Y(:, :)
+        end function diff_2_${k1}$
+        #:endfor
+    end interface diff
```

Comment:
Should this be extended to all dimensions supported by Fortran (up to 7 or to 15)?

Reply:
Fortran does not seem to be as easy to create an array interface for any `rank` like dynamic languages python and matlab. From a cost point of view, it seems more appropriate to only support `rank-1/rank-2`.

If there is a higher demand for `rank` in the future, maybe we can expand it?

Reply:
The `stdlib_stats` routines use some fancy fypp to enable treating any rank and `dim`. See around [this part of the code](https://github.com/fortran-lang/stdlib/blob/03b1695fdd514c5ad9dc5e1292b6eda178d28ab4/src/stdlib_stats_median.fypp#L93)

Reply:
Yes, using fypp can achieve higher-level array rank support, but I don’t know whether we need higher-level rank support at this stage. Maybe we can be conservative?